File: benchmarks/public_suite/cpu_matrix/matrix_blocked.go
package main

import (
	"fmt"
	"os"
	"strconv"
	"time"
)

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

func matrixMultBlocked(n int) float64 {
	a := make([]float64, n*n)
	b := make([]float64, n*n)
	c := make([]float64, n*n)
	for i := 0; i < n*n; i++ {
		a[i] = 1.0
		b[i] = 2.0
	}

	block := 32
	for i0 := 0; i0 < n; i0 += block {
		imax := min(i0+block, n)
		for k0 := 0; k0 < n; k0 += block {
			kmax := min(k0+block, n)
			for j0 := 0; j0 < n; j0 += block {
				jmax := min(j0+block, n)

				for i := i0; i < imax; i++ {
					for k := k0; k < kmax; k++ {
						aik := a[i*n+k]
						for j := j0; j < jmax; j++ {
							c[i*n+j] += aik * b[k*n+j]
						}
					}
				}
			}
		}
	}

	return c[0] + c[n*n-1]
}

func main() {
	n := 500
	if len(os.Args) > 1 {
		if val, err := strconv.Atoi(os.Args[1]); err == nil {
			n = val
		}
	}

	t1 := time.Now()
	checksum := matrixMultBlocked(n)
	ms := float64(time.Since(t1).Nanoseconds()) / 1e6

	fmt.Printf("{\"lang\": \"Go (Blocked)\", \"test\": \"Matrix 500x500\", \"checksum\": %.1f, \"time_ms\": %.3f}\n", checksum, ms)
}

File: benchmarks/public_suite/cpu_matrix/matrix_blocked.rs
use std::env;
use std::time::Instant;

fn matrix_mult_blocked(n: usize) -> f64 {
    let a = vec![1.0f64; n * n];
    let b = vec![2.0f64; n * n];
    let mut c = vec![0.0f64; n * n];

    let block = 32;
    for i0 in (0..n).step_by(block) {
        let imax = (i0 + block).min(n);
        for k0 in (0..n).step_by(block) {
            let kmax = (k0 + block).min(n);
            for j0 in (0..n).step_by(block) {
                let jmax = (j0 + block).min(n);

                for i in i0..imax {
                    for k in k0..kmax {
                        let aik = a[i * n + k];
                        for j in j0..jmax {
                            c[i * n + j] += aik * b[k * n + j];
                        }
                    }
                }
            }
        }
    }

    c[0] + c[n * n - 1]
}

fn main() {
    let args: Vec<String> = env::args().collect();
    let n: usize = if args.len() > 1 { args[1].parse().unwrap_or(500) } else { 500 };

    let t1 = Instant::now();
    let checksum = matrix_mult_blocked(n);
    let ms = t1.elapsed().as_secs_f64() * 1000.0;

    println!("{{\"lang\": \"Rust (Blocked)\", \"test\": \"Matrix 500x500\", \"checksum\": {}, \"time_ms\": {:.3}}}", checksum, ms);
}
